### PR TITLE
fix Issue 17425 - add __traits(getParameterStorageClasses, f, i)

### PR DESCRIPTION
**Pull request by WalterBright**
The point of this is so that `std.traits.ParameterStorageClassTuple()` can be implemented without needing to study the name mangling.

### CODE REVIEW

**Comment by WalterBright** [comment]
Blocking https://github.com/dlang/phobos/pull/5427

**Comment by UplinkCoder** [comment]
@WalterBright there is some kind of error on osx32
I am not sure if it is related.

**Comment by WalterBright** [comment]
> I am not sure if it is related.

Apparently another heisenbug in the testing scripts.

**Review by UplinkCoder — APPROVED** [review]
Looks fine to me.

File: changelog/changelog17425.dd [hunk 1]
```diff
@@ -0,0 +1,18 @@
+add __traits(getParameterStorageClasses, f, i)
+
+$(LINK2 https://issues.dlang.org/show_bug.cgi?id=17425, Bugzilla 17425)
```
Comment (rainers) [inline changelog/changelog17425.dd]: The changelog entry seems a bit terse. I'd use at least more descriptive variable names. Also, linking to the documentation instead of bugzilla might be more helpful.

File: src/ddmd/traits.d [hunk 2]
```diff
@@ -968,6 +969,109 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
         auto se = new StringExp(e.loc, cast(char*)style);
         return se.semantic(sc);
     }
+    if (e.ident == Id.getParameterStorageClasses)
+    {
+        /* Accept a function symbol or a type, followed by a parameter index.
```
Comment (rainers) [inline src/ddmd/traits.d]: How about putting this code into a function? This improves readable and avoids making this function even larger (dmd is rather slow optimizing large functions).

File: src/ddmd/traits.d [hunk 3]
```diff
@@ -968,6 +969,109 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
...
+        // Set stc to storage class of the ith parameter
+        auto ex = isExpression((*e.args)[1]);
+        if (!ex)
+        {
+            e.error("expression expected as second argument of `__traits(getParameterStorageClasses, %s, %s)`",
+                o.toChars(), o1.toChars());
+            return new ErrorExp();
+        }
+        ex = ex.ctfeInterpret();
+        auto ii = ex.toUInteger();
```
Comment (rainers) [inline src/ddmd/traits.d]: This doesn't propagate errors if the expression does not evaluate to an integer. That is not checked in other uses of toInteger/toUInteger elsewhere, too, though. Maybe add test `pragma(msg, __traits(getParameterStorageClasses, foo, "1"));`?

**Review by rainers — APPROVED** [review]
Just a few nits, otherwise LGTM.

File: changelog/changelog17425.dd (hunk 1 above)
Comment (wilzbach) [inline changelog/changelog17425.dd]: FYI the auto-tester allows to preview this since some time (though I am not sure why this isn't part of the diff, CC @CyberShadow)
In any case the link is e.g. http://dtest.dlang.io/artifact/website-72ba79b73f635fb08bdfe69d6826833c1c4c2ee0-a9052bf749bf29e4ae70eb23d51613de/web/changelog/2.075.0_pre.html#changelog17425 (the _pre file in the changelog folder).
FWIW there also seems to be a parenthesis problem in another changelog file..

File: src/ddmd/traits.d (hunk 2 above)
Comment (WalterBright) [inline src/ddmd/traits.d]: I leave refactorings to subsequent PRs, because they distract from what is being changed here.

File: src/ddmd/traits.d (hunk 3 above)
Comment (WalterBright) [inline src/ddmd/traits.d]: That's a generic problem with all those functions, and fixing them should be in a separate PR.

File: changelog/changelog17425.dd (hunk 1 above)
Comment (WalterBright) [inline changelog/changelog17425.dd]: I didn't link to the documentation because that PR has not been pulled.

**Comment by CyberShadow** [comment]
> though I am not sure why this isn't part of the diff, CC @CyberShadow

Oops, diff was eaten by an overly voracious disk cleanup script. Fixed for future cleanups.
